app/app.go

```diff
@@ -1294,32 +1294,49 @@ func (d *Deployments) GetDeploymentForDeviceWithCurrent(ctx context.Context, dev
 		return nil, nil
 	}
 
-	if installed.ArtifactName != "" && deployment.ArtifactName == installed.ArtifactName {
-		// if the device reported same artifact name as the one in the device deployment,
-		// and this is a new device deployment - indicated by device deployment
-		// status "pending",
-		// pretend there is no deployment for this device, but update
-		// its status to already installed first
-		if deviceDeployment.Status == model.DeviceDeploymentStatusPending {
+	if deviceDeployment.Request != nil &&
+		(deviceDeployment.Request.DeviceType != installed.DeviceType ||
+			deviceDeployment.Request.ArtifactName != installed.ArtifactName) {
+		// the device reported different device type and/or artifact name
+		// during the update process, which should never happen;
+		// mark deployment for this device as failed to force client to rollback
+		l.Errorf(
+			"Device with id %s reported new data during update process - "+
+				"device type: %s, artifact name: %s "+
+				"old data - device type %s, artifact name: %s",
+			deviceID, installed.DeviceType, installed.ArtifactName,
+			deviceDeployment.Request.DeviceType, deviceDeployment.Request.ArtifactName)
 
-			if err := d.UpdateDeviceDeploymentStatus(ctx, deviceDeployment.DeploymentId, deviceID,
-				model.DeviceDeploymentState{
-					Status: model.DeviceDeploymentStatusAlreadyInst,
-				}); err != nil {
+		if err := d.UpdateDeviceDeploymentStatus(ctx, deviceDeployment.DeploymentId, deviceID,
+			model.DeviceDeploymentState{
+				Status: model.DeviceDeploymentStatusFailure,
+			}); err != nil {
 
-				return nil, errors.Wrap(err, "Failed to update deployment status")
-			}
+			return nil, errors.Wrap(err, "Failed to update deployment status")
+		}
+		return nil, nil
```

Comment:
Should there be an error definition for this one?
I think this deserves a HTTP/409 Conflict response - this will also make the client abort the deployment.

Reply:
I'm not sure what client will do exactly; I guess it will at least retry in case of error, right? there is no special handling for 409 error in the client

Reply:
I believe it will fail the deployment if it receives an unexpected status code.

Reply:
done